### ISSUE TITLE
Implement a cache to not compile the Wasm module everytime

### ISSUE THREAD

**Issue by Hywan** [wasmerio/php-ext-wasm#18]
The idea is basically to add a cache, so that we don't have to validate & compile etc. the whole module everytime. It will improve grately the execution speed.

There is no API in `wasmer-runtime-c-api` yet. I need to work on that. Maybe we will implement that outside the `wasmer-runtime-c-api` crate.

**Comment by xtuc** [comment]
I was about to open a issue to ask you a question; could you please clarify the lifetime of a WebAssembly Instance/Module?

I would like to avoid to instantiate the WebAssembly module for each request. You are mentioning a cache for the module, I'm assuming that's to avoid running possible side-effects at instantiation, right?

However, I think it would be useful to explicitly ask to keep the instance in cache too (forgive my ignorance about PHP extensions). Would that be possible?

**Comment by Hywan** [comment]
So far, the lifetime of a module is constrainted by the lifetime of a PHP execution. It starts when your run `wasm_new_instance`, and it is destroyed when the instance is destroyed (so either with `unset` or by waiting the end of the execution).

> I would like to avoid to instantiate the WebAssembly module for each request

This issue is here to only avoid the compilation step. The cache will contain a module then. I'm likely to use https://github.com/wasmerio/wasmer/blob/master/lib/runtime-core/src/cache.rs.

> You are mentioning a cache for the module, I'm assuming that's to avoid running possible side-effects at instantiation, right?

Do you mean caching the module instead of an instance? Yes. I don't want to cache the instance. That would be a terrible from my point of view, since it comes with the memory pages, tables etc.

> However, I think it would be useful to explicitly ask to keep the instance in cache too (forgive my ignorance about PHP extensions). Would that be possible?

It is, but I don't want to address that yet. I want to first cache the module, and see how it impacts the performance in general.

It's trickier to cache an instance. And it depends of the usecases. Let's see! … or do you have a specific usecase in mind?

**Comment by xtuc** [comment]
> It's trickier to cache an instance. And it depends of the usecases. Let's see! … or do you have a specific usecase in mind?

Well I like PHP but here I have an occassion to avoid it. I don't have a specific use case in mind, but everything that would retain state. For intance, that could replace memcached with an app specific cache!

I agree that this shouldn't be the default caching strategy, but an explicit API would be great. I believe it's out-of-scope for now, just saying it would be nice to have.

**Comment by Hywan** [comment]
I'm starting drafting it. More to come tomorow :-). I will probably ping you in the PR, if you're agree?

**Comment by xtuc** [comment]
Sure, I'm happy to help.

**Comment by Hywan** [comment]
Waiting on https://github.com/wasmerio/wasmer/pull/271.
